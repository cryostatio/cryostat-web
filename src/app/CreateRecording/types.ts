/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { EventTemplate, KeyValue } from '@app/Shared/Services/api.types';
import { ValidatedOptions } from '@patternfly/react-core';

export type EventTemplateIdentifier = Pick<EventTemplate, 'name' | 'type'>;

interface _FormBaseData {
  name: string;
  template?: EventTemplateIdentifier;
  labels: KeyValue[];
  continuous: boolean;
  archiveOnStop: boolean;
  restart: boolean;
  duration: number;
  durationUnit: number;
  skipDurationCheck?: boolean;
  maxAge: number;
  maxAgeUnit: number;
  maxSize: number;
  maxSizeUnit: number;
  toDisk: boolean;
}

interface _FormValidationData {
  nameValid: ValidatedOptions;
  labelsValid: ValidatedOptions;
  durationValid: ValidatedOptions;
}

export type CustomRecordingFormData = _FormBaseData & _FormValidationData;
