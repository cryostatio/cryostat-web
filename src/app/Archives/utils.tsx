/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { RecordingDirectory, Target } from '@app/Shared/Services/api.types';

export const includesDirectory = (arr: RecordingDirectory[], dir: RecordingDirectory): boolean => {
  return arr.some((t) => t.connectUrl === dir.connectUrl);
};

export const indexOfDirectory = (arr: RecordingDirectory[], dir: RecordingDirectory): number => {
  let index = -1;
  arr.forEach((d, idx) => {
    if (d.connectUrl === dir.connectUrl) {
      index = idx;
    }
  });
  return index;
};

export const getTargetFromDirectory = (dir: RecordingDirectory): Target => {
  return {
    connectUrl: dir.connectUrl,
    alias: dir.jvmId,
    labels: [],
    annotations: {
      cryostat: [],
      platform: [],
    },
  };
};
