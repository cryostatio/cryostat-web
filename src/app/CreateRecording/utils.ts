/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { TemplateType } from '@app/Shared/Services/api.types';
import { EventTemplateIdentifier } from './types';

export const RecordingNamePattern = /^[\w_]+$/;

export const DurationPattern = /^[1-9][0-9]*$/;

export const isRecordingNameValid = (name: string) => RecordingNamePattern.test(name);

export const isDurationValid = (duration: number) => DurationPattern.test(`${duration}`);

export const templateFromEventSpecifier = (eventSpecifier: string): EventTemplateIdentifier | undefined => {
  const regex = /^template=([a-zA-Z0-9]+)(?:,type=([a-zA-Z0-9]+))?$/im;
  if (eventSpecifier && regex.test(eventSpecifier)) {
    const parts = regex.exec(eventSpecifier);
    if (parts) {
      return {
        name: parts[1],
        type: parts[2] as TemplateType,
      };
    }
  }
  return undefined;
};
