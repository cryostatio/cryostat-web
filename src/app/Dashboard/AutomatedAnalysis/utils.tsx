/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { AnalysisResult, Evaluation } from '@app/Shared/Services/api.types';
import { Stack, StackItem, Label, Title, Text } from '@patternfly/react-core';
import _ from 'lodash';

export const transformAADescription = (result: AnalysisResult): JSX.Element => {
  const format = (s: Evaluation): JSX.Element => {
    if (typeof s === 'string') {
      return <Text>{s}</Text>;
    }
    if (Array.isArray(s)) {
      return (
        <Stack>
          {s.map((e) => (
            <StackItem key={e.setting}>
              <Title headingLevel={'h6'}>{e.setting}</Title>
              <Label>
                {e.name}={e.value}
              </Label>
            </StackItem>
          ))}
        </Stack>
      );
    }
    throw `Unrecognized item: ${s}`;
  };
  return (
    <div>
      {Object.entries(result.evaluation || {}).map(([k, v]) =>
        v && v.length ? (
          <div key={k}>
            <span>
              <Title headingLevel={'h5'}>{_.capitalize(k)}</Title>
              {format(result.evaluation[k])}
            </span>
            <br />
          </div>
        ) : (
          <div key={k}></div>
        ),
      )}
    </div>
  );
};
