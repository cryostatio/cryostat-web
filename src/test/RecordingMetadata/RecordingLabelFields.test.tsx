/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { RecordingLabelFields, RecordingLabelFieldsProps } from '@app/RecordingMetadata/RecordingLabelFields';
import { KeyValue } from '@app/Shared/Services/api.types';
import { ValidatedOptions } from '@patternfly/react-core';
import '@testing-library/jest-dom';
import * as tlr from '@testing-library/react';
import { cleanup, screen } from '@testing-library/react';
import { render, renderSnapshot } from '../utils';

const mockUploadedRecordingLabels = {
  someUploaded: 'someUploadedValue',
};
const mockMetadataFileName = 'mock.metadata.json';
const mockMetadataFile = new File(
  [JSON.stringify({ labels: { ...mockUploadedRecordingLabels } })],
  mockMetadataFileName,
  { type: 'json' },
);
mockMetadataFile.text = jest.fn(
  () => new Promise((resolve, _) => resolve(JSON.stringify({ labels: { ...mockUploadedRecordingLabels } }))),
);

describe('<RecordingLabelFields />', () => {
  // RecordingLabelFields component modifies labels in-place, so we need to reinitialize mocks
  // after every tests.
  let mockLabels: KeyValue[];
  let mockValid: ValidatedOptions;
  let mockProps: RecordingLabelFieldsProps;
  let mockLabel1: KeyValue;
  let mockLabel2: KeyValue;
  let mockEmptyLabel: KeyValue;

  afterEach(cleanup);

  beforeEach(() => {
    mockLabel1 = {
      key: 'someLabel',
      value: 'someValue',
    } as KeyValue;
    mockLabel2 = {
      key: 'anotherLabel',
      value: 'anotherValue',
    } as KeyValue;
    mockEmptyLabel = {
      key: '',
      value: '',
    } as KeyValue;
    mockLabels = [mockLabel1, mockLabel2];
    mockValid = ValidatedOptions.default;
    mockProps = {
      labels: mockLabels,
      setLabels: jest.fn((labels: KeyValue[]) => (mockLabels = labels.slice())),
      setValid: jest.fn((state: ValidatedOptions) => (mockValid = state)),
    };
  });

  it('renders correctly', async () => {
    const tree = await renderSnapshot({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <RecordingLabelFields {...mockProps} />,
          },
        ],
      },
    });
    expect(tree).toMatchSnapshot();
  });

  it('displays all labels in form fields', async () => {
    render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <RecordingLabelFields {...mockProps} />,
          },
        ],
      },
    });

    const inputs = [
      screen.getByDisplayValue('someLabel'),
      screen.getByDisplayValue('someValue'),
      screen.getByDisplayValue('anotherLabel'),
      screen.getByDisplayValue('anotherValue'),
    ];

    inputs.forEach((input) => {
      expect(input).toBeInTheDocument();
      expect(input).toBeVisible();
    });

    const addLabelButton = screen.getByRole('button', { name: 'Add Label' });
    expect(addLabelButton).toBeInTheDocument();
    expect(addLabelButton).toBeVisible();

    expect(mockProps.setValid).toHaveBeenCalledTimes(1);
    expect(mockProps.setValid).toHaveBeenCalledWith(ValidatedOptions.success);
  });

  it('updates the label key when entering text into the Key input', async () => {
    const { user } = render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <RecordingLabelFields {...mockProps} />,
          },
        ],
      },
    });

    const labelKeyInput = screen.getAllByLabelText('Label Key')[0] as HTMLInputElement;
    expect(labelKeyInput).toBeInTheDocument();
    expect(labelKeyInput).toBeVisible();

    labelKeyInput.setSelectionRange(0, mockProps.labels[0].key.length);
    labelKeyInput.focus();
    await user.paste('someEditedKey');

    expect(mockProps.labels[0].key).toBe('someEditedKey');
  });

  it('updates the label value when entering text into the Value input', async () => {
    const { user } = render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <RecordingLabelFields {...mockProps} />,
          },
        ],
      },
    });

    const labelValueInput = screen.getAllByLabelText('Label Value')[0] as HTMLInputElement;
    expect(labelValueInput).toBeInTheDocument();
    expect(labelValueInput).toBeVisible();

    labelValueInput.setSelectionRange(0, mockProps.labels[0].value.length);
    labelValueInput.focus();
    await user.paste('someEditedValue');

    expect(mockProps.labels[0].value).toBe('someEditedValue');
  });

  it('validates labels on initial render', async () => {
    render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <RecordingLabelFields {...mockProps} />,
          },
        ],
      },
    });

    expect(mockProps.setValid).toHaveBeenCalledTimes(1);
    expect(mockProps.setValid).toHaveBeenCalledWith(ValidatedOptions.success);
    expect(mockValid).toBe(ValidatedOptions.success);
  });

  it('adds a label when Add Label is clicked', async () => {
    const { user } = render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <RecordingLabelFields {...mockProps} />,
          },
        ],
      },
    });

    expect(screen.getAllByLabelText('Label Key').length).toBe(2);
    expect(screen.getAllByLabelText('Label Value').length).toBe(2);

    await user.click(screen.getByText('Add Label'));

    expect(mockProps.setLabels).toHaveBeenCalledTimes(1);
    expect(mockProps.setLabels).toHaveBeenCalledWith([mockLabel1, mockLabel2, mockEmptyLabel]);
  });

  it('removes the correct label when Delete button is clicked', async () => {
    const { user } = render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <RecordingLabelFields {...mockProps} />,
          },
        ],
      },
    });

    expect(screen.getAllByLabelText('Label Key').length).toBe(2);
    expect(screen.getAllByLabelText('Label Value').length).toBe(2);

    await user.click(screen.getAllByLabelText('Remove Label')[0]);

    expect(mockProps.setLabels).toHaveBeenCalledTimes(1);
    expect(mockProps.setLabels).toHaveBeenCalledWith([mockLabel2]);
  });

  it('validates the label key when valid', async () => {
    render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <RecordingLabelFields {...mockProps} />,
          },
        ],
      },
    });

    expect(mockValid).toBe(ValidatedOptions.success);

    screen.getAllByLabelText('Label Key').forEach((element) => {
      expect(element).toBeInTheDocument();
      expect(element).toBeVisible();
      expect(element.classList.contains('pf-m-success')).toBe(true);
    });
  });

  it('validates the label value when valid', async () => {
    render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <RecordingLabelFields {...mockProps} />,
          },
        ],
      },
    });

    expect(mockValid).toBe(ValidatedOptions.success);

    screen.getAllByLabelText('Label Value').forEach((element) => {
      expect(element).toBeInTheDocument();
      expect(element).toBeVisible();
      expect(element.classList.contains('pf-m-success')).toBe(true);
    });
  });

  it('invalidates form when the label key is invalid', async () => {
    const invalidLabels = [{ key: 'label with whitespace', value: 'someValue' }];

    render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <RecordingLabelFields {...mockProps} labels={invalidLabels} />,
          },
        ],
      },
    });

    expect(mockValid).toBe(ValidatedOptions.error);
  });

  it('invalidates form when the label value is invalid', async () => {
    const invalidLabels = [{ key: 'someLabel', value: 'value with whitespace' }];

    render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <RecordingLabelFields {...mockProps} labels={invalidLabels} />,
          },
        ],
      },
    });

    expect(mockValid).toBe(ValidatedOptions.error);
  });

  it('shows error text when the label key is invalid', async () => {
    const invalidLabels = [{ key: 'label with whitespace', value: 'someValue' }];

    render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <RecordingLabelFields {...mockProps} labels={invalidLabels} />,
          },
        ],
      },
    });

    expect(mockValid).toBe(ValidatedOptions.error);

    const errorText = await screen.findByText('Keys must be unique. Labels should not contain empty spaces.');
    expect(errorText).toBeInTheDocument();
    expect(errorText).toBeVisible();
  });

  it('shows error text when the label value is invalid', async () => {
    const invalidLabels = [{ key: 'someLabel', value: 'value with whitespace' }];

    render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <RecordingLabelFields {...mockProps} labels={invalidLabels} />,
          },
        ],
      },
    });

    expect(mockValid).toBe(ValidatedOptions.error);

    const errorText = await screen.findByText('Keys must be unique. Labels should not contain empty spaces.');
    expect(errorText).toBeInTheDocument();
    expect(errorText).toBeVisible();
  });

  it('shows upload button when upload is enabled ', async () => {
    mockProps.isUploadable = true;

    render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <RecordingLabelFields {...mockProps} />,
          },
        ],
      },
    });

    const uploadButton = screen.getByRole('button', { name: 'Upload Labels' });
    expect(uploadButton).toBeInTheDocument();
    expect(uploadButton).toBeVisible();
  });

  it('updates label list when upload is enabled and upload is submitted', async () => {
    mockProps.isUploadable = true;

    const { user } = render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: <RecordingLabelFields {...mockProps} />,
          },
        ],
      },
    });

    const uploadButton = screen.getByRole('button', { name: 'Upload Labels' });
    expect(uploadButton).toBeInTheDocument();
    expect(uploadButton).toBeVisible();

    await user.click(uploadButton);

    const labelUploadInput = document.querySelector("input[accept='.json'][type='file']") as HTMLInputElement;
    expect(labelUploadInput).toBeInTheDocument();

    await tlr.act(async () => {
      await user.upload(labelUploadInput, mockMetadataFile);
    });

    expect(labelUploadInput.files).not.toBe(null);
    expect(labelUploadInput.files?.item(0)).toStrictEqual(mockMetadataFile);

    expect(mockProps.setLabels).toHaveBeenCalledTimes(1);
    expect(mockProps.setLabels).toHaveBeenCalledWith([
      mockLabel1,
      mockLabel2,
      { key: 'someUploaded', value: 'someUploadedValue' },
    ]);
  });
});
