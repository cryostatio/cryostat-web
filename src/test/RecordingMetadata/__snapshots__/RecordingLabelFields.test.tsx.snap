// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RecordingLabelFields /> renders correctly 1`] = `
Array [
  <button
    aria-disabled={false}
    aria-label="Add Label"
    className="pf-v5-c-button pf-m-link"
    data-ouia-component-id="OUIA-Generated-Button-link-1"
    data-ouia-component-type="PF5/Button"
    data-ouia-safe={true}
    disabled={false}
    onClick={[Function]}
    role={null}
    type="button"
  >
    <span
      className="pf-v5-c-button__icon pf-m-start"
    >
      <svg
        aria-hidden={true}
        aria-labelledby={null}
        className="pf-v5-svg"
        fill="currentColor"
        height="1em"
        role="img"
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm144 276c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92h-92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z"
        />
      </svg>
    </span>
    Add Label
  </button>,
  <div
    className="pf-v5-l-split pf-m-gutter"
  >
    <div
      className="pf-v5-l-split__item pf-m-fill"
    >
      <span
        className="pf-v5-c-form-control pf-m-success"
      >
        <input
          aria-describedby="label-key-input-helper"
          aria-invalid={false}
          aria-label="Label Key"
          data-ouia-component-id="OUIA-Generated-TextInputBase-1"
          data-ouia-component-type="PF5/TextInput"
          data-ouia-safe={true}
          disabled={false}
          id="label-key-input"
          name="label-key-input"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={true}
          type="text"
          value="someLabel"
        />
        <span
          className="pf-v5-c-form-control__utilities"
        >
          <span
            className="pf-v5-c-form-control__icon pf-m-status"
          >
            <svg
              aria-hidden={true}
              aria-labelledby={null}
              className="pf-v5-svg"
              fill="currentColor"
              height="1em"
              role="img"
              viewBox="0 0 512 512"
              width="1em"
            >
              <path
                d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
              />
            </svg>
          </span>
        </span>
      </span>
      <p
        className=""
        data-ouia-component-id="OUIA-Generated-Text-1"
        data-ouia-component-type="PF5/Text"
        data-ouia-safe={true}
        data-pf-content={true}
      >
        Key
      </p>
      <div
        className="pf-v5-c-form__helper-text"
      >
        <div
          className="pf-v5-c-helper-text"
          id="label-error-text"
        >
          <div
            className="pf-v5-c-helper-text__item pf-m-error"
          >
            <span
              className="pf-v5-c-helper-text__item-text"
            >
              Keys must be unique. Labels should not contain empty spaces.
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      className="pf-v5-l-split__item pf-m-fill"
    >
      <span
        className="pf-v5-c-form-control pf-m-success"
      >
        <input
          aria-describedby="label-value-input-helper"
          aria-invalid={false}
          aria-label="Label Value"
          data-ouia-component-id="OUIA-Generated-TextInputBase-2"
          data-ouia-component-type="PF5/TextInput"
          data-ouia-safe={true}
          disabled={false}
          id="label-value-input"
          name="label-value-input"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={true}
          type="text"
          value="someValue"
        />
        <span
          className="pf-v5-c-form-control__utilities"
        >
          <span
            className="pf-v5-c-form-control__icon pf-m-status"
          >
            <svg
              aria-hidden={true}
              aria-labelledby={null}
              className="pf-v5-svg"
              fill="currentColor"
              height="1em"
              role="img"
              viewBox="0 0 512 512"
              width="1em"
            >
              <path
                d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
              />
            </svg>
          </span>
        </span>
      </span>
      <p
        className=""
        data-ouia-component-id="OUIA-Generated-Text-2"
        data-ouia-component-type="PF5/Text"
        data-ouia-safe={true}
        data-pf-content={true}
      >
        Value
      </p>
    </div>
    <div
      className="pf-v5-l-split__item"
    >
      <button
        aria-disabled={false}
        aria-label="Remove Label"
        className="pf-v5-c-button pf-m-link"
        data-ouia-component-id="OUIA-Generated-Button-link-2"
        data-ouia-component-type="PF5/Button"
        data-ouia-safe={true}
        disabled={false}
        onClick={[Function]}
        role={null}
        type="button"
      >
        <span
          className="pf-v5-c-button__icon pf-m-start"
        >
          <span
            className="pf-v5-c-icon pf-m-sm"
          >
            <span
              className="pf-v5-c-icon__content"
            >
              <svg
                aria-hidden={true}
                aria-labelledby={null}
                className="pf-v5-svg"
                color="gray"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 730 1024"
                width="1em"
              >
                <path
                  d="M725.499315,767.757345 L469.242169,511.500499 L725.499315,255.243653 C729.196254,251.545719 731.096195,247.248119 731,242.4498 C731,237.653479 728.997365,233.35588 725.299427,229.558002 L647.542773,151.701495 C643.743891,147.903617 639.446287,146 634.64996,146 C629.852635,146 625.55503,147.803673 621.857092,151.500607 L365.5,407.857398 L109.241857,151.500607 C105.544917,147.803673 101.246314,145.904733 96.4499876,146 C91.6526619,146.104622 87.3550577,148.002562 83.5571748,151.701495 L5.7005771,229.457058 C1.90269428,233.255936 0,237.553535 0,242.350855 C0,247.148175 1.80175055,251.444775 5.50068853,255.143709 L261.857779,511.500499 L5.50068853,767.757345 C1.80175055,771.454279 -0.0961914123,775.752878 0,779.750645 C0.103697157,784.547965 2.00263857,788.845564 5.7005771,792.642443 L83.4572306,871.299502 C87.2541139,875.096381 91.5527176,877 96.3500433,877 C101.147369,877 105.444973,875.197325 109.142912,871.499391 L365.400058,615.241545 L621.657203,871.499391 C625.355142,875.197325 629.652746,877.095265 634.449072,877 C639.246398,877 643.545002,874.997437 647.342885,871.299502 L725.099538,793.54294 C728.896421,789.745062 730.796362,785.446463 730.796362,780.649143 C731.096195,775.752878 729.196254,771.454279 725.499315,767.757345"
                />
              </svg>
               
            </span>
          </span>
        </span>
      </button>
    </div>
  </div>,
  <div
    className="pf-v5-l-split pf-m-gutter"
  >
    <div
      className="pf-v5-l-split__item pf-m-fill"
    >
      <span
        className="pf-v5-c-form-control pf-m-success"
      >
        <input
          aria-describedby="label-key-input-helper"
          aria-invalid={false}
          aria-label="Label Key"
          data-ouia-component-id="OUIA-Generated-TextInputBase-3"
          data-ouia-component-type="PF5/TextInput"
          data-ouia-safe={true}
          disabled={false}
          id="label-key-input"
          name="label-key-input"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={true}
          type="text"
          value="anotherLabel"
        />
        <span
          className="pf-v5-c-form-control__utilities"
        >
          <span
            className="pf-v5-c-form-control__icon pf-m-status"
          >
            <svg
              aria-hidden={true}
              aria-labelledby={null}
              className="pf-v5-svg"
              fill="currentColor"
              height="1em"
              role="img"
              viewBox="0 0 512 512"
              width="1em"
            >
              <path
                d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
              />
            </svg>
          </span>
        </span>
      </span>
      <p
        className=""
        data-ouia-component-id="OUIA-Generated-Text-3"
        data-ouia-component-type="PF5/Text"
        data-ouia-safe={true}
        data-pf-content={true}
      >
        Key
      </p>
      <div
        className="pf-v5-c-form__helper-text"
      >
        <div
          className="pf-v5-c-helper-text"
          id="label-error-text"
        >
          <div
            className="pf-v5-c-helper-text__item pf-m-error"
          >
            <span
              className="pf-v5-c-helper-text__item-text"
            >
              Keys must be unique. Labels should not contain empty spaces.
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      className="pf-v5-l-split__item pf-m-fill"
    >
      <span
        className="pf-v5-c-form-control pf-m-success"
      >
        <input
          aria-describedby="label-value-input-helper"
          aria-invalid={false}
          aria-label="Label Value"
          data-ouia-component-id="OUIA-Generated-TextInputBase-4"
          data-ouia-component-type="PF5/TextInput"
          data-ouia-safe={true}
          disabled={false}
          id="label-value-input"
          name="label-value-input"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={true}
          type="text"
          value="anotherValue"
        />
        <span
          className="pf-v5-c-form-control__utilities"
        >
          <span
            className="pf-v5-c-form-control__icon pf-m-status"
          >
            <svg
              aria-hidden={true}
              aria-labelledby={null}
              className="pf-v5-svg"
              fill="currentColor"
              height="1em"
              role="img"
              viewBox="0 0 512 512"
              width="1em"
            >
              <path
                d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
              />
            </svg>
          </span>
        </span>
      </span>
      <p
        className=""
        data-ouia-component-id="OUIA-Generated-Text-4"
        data-ouia-component-type="PF5/Text"
        data-ouia-safe={true}
        data-pf-content={true}
      >
        Value
      </p>
    </div>
    <div
      className="pf-v5-l-split__item"
    >
      <button
        aria-disabled={false}
        aria-label="Remove Label"
        className="pf-v5-c-button pf-m-link"
        data-ouia-component-id="OUIA-Generated-Button-link-3"
        data-ouia-component-type="PF5/Button"
        data-ouia-safe={true}
        disabled={false}
        onClick={[Function]}
        role={null}
        type="button"
      >
        <span
          className="pf-v5-c-button__icon pf-m-start"
        >
          <span
            className="pf-v5-c-icon pf-m-sm"
          >
            <span
              className="pf-v5-c-icon__content"
            >
              <svg
                aria-hidden={true}
                aria-labelledby={null}
                className="pf-v5-svg"
                color="gray"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 730 1024"
                width="1em"
              >
                <path
                  d="M725.499315,767.757345 L469.242169,511.500499 L725.499315,255.243653 C729.196254,251.545719 731.096195,247.248119 731,242.4498 C731,237.653479 728.997365,233.35588 725.299427,229.558002 L647.542773,151.701495 C643.743891,147.903617 639.446287,146 634.64996,146 C629.852635,146 625.55503,147.803673 621.857092,151.500607 L365.5,407.857398 L109.241857,151.500607 C105.544917,147.803673 101.246314,145.904733 96.4499876,146 C91.6526619,146.104622 87.3550577,148.002562 83.5571748,151.701495 L5.7005771,229.457058 C1.90269428,233.255936 0,237.553535 0,242.350855 C0,247.148175 1.80175055,251.444775 5.50068853,255.143709 L261.857779,511.500499 L5.50068853,767.757345 C1.80175055,771.454279 -0.0961914123,775.752878 0,779.750645 C0.103697157,784.547965 2.00263857,788.845564 5.7005771,792.642443 L83.4572306,871.299502 C87.2541139,875.096381 91.5527176,877 96.3500433,877 C101.147369,877 105.444973,875.197325 109.142912,871.499391 L365.400058,615.241545 L621.657203,871.499391 C625.355142,875.197325 629.652746,877.095265 634.449072,877 C639.246398,877 643.545002,874.997437 647.342885,871.299502 L725.099538,793.54294 C728.896421,789.745062 730.796362,785.446463 730.796362,780.649143 C731.096195,775.752878 729.196254,771.454279 725.499315,767.757345"
                />
              </svg>
               
            </span>
          </span>
        </span>
      </button>
    </div>
  </div>,
]
`;
